pkg/overlay/server.go

```diff
@@ -74,52 +81,76 @@ func (server *Server) FindStorageNodes(ctx context.Context, req *pb.FindStorageN
 	defer mon.Task()(&ctx)(&err)
 
 	opts := req.GetOpts()
-	maxNodes := req.GetMaxNodes()
-	if maxNodes <= 0 {
-		maxNodes = opts.GetAmount()
+	// todo(nat): change name from MaxNodes to minRequestedNodes or similar
+	requiredReputableNodes := req.GetMaxNodes()
+	if requiredReputableNodes <= 0 {
+		requiredReputableNodes = opts.GetAmount()
 	}
 
 	excluded := opts.ExcludedNodes
 	restrictions := opts.GetRestrictions()
-	reputation := server.nodeStats
+	usedAddrs := make(map[string]bool)
+
+	newStartID := req.Start
+	requiredNewNodes := int64(float64(requiredReputableNodes) * server.newNodePercentage)
+	var allNewNodes []*pb.Node
+
+	reputableStartID := req.Start
+	var allReputableNodes []*pb.Node
 
-	var startID storj.NodeID
-	result := []*pb.Node{}
 	for {
-		var nodes []*pb.Node
-		nodes, startID, err = server.populate(ctx, req.Start, maxNodes, restrictions, reputation, excluded)
+		var reputableNodes []*pb.Node
+		var err error
+		reputableNodes, reputableStartID, err = server.getReputableNodes(ctx, reputableStartID, requiredReputableNodes, restrictions, excluded)
 		if err != nil {
 			return nil, Error.Wrap(err)
 		}
 
-		resultNodes := []*pb.Node{}
-		usedAddrs := make(map[string]bool)
-		for _, n := range nodes {
-			addr := n.Address.GetAddress()
-			excluded = append(excluded, n.Id) // exclude all nodes on next iteration
-			if !usedAddrs[addr] {
-				resultNodes = append(resultNodes, n)
-				usedAddrs[addr] = true
-			}
-		}
-		if len(resultNodes) <= 0 {
+		reputableNodes, excluded, usedAddrs = server.filterNodes(ctx, reputableNodes, requiredReputableNodes, usedAddrs, excluded)
+
+		allReputableNodes = append(allReputableNodes, reputableNodes...)
+
+		// this means we've exhausted the overlay cache looking for reputable nodes or we have enough
+		if reputableStartID == (storj.NodeID{}) || int64(len(allReputableNodes)) >= requiredReputableNodes {
```

Comment:
nodeid has method `IsEmpty`.

Reply:
I see `IsZero` not `IsEmpty`, is that doing the same thing?